src/default.rs

```diff
@@ -20,27 +20,58 @@ thread_local! {
 /// Pins the current thread.
 #[inline]
 pub fn pin() -> Guard {
-    // FIXME(jeehoonkang): thread-local storage may be destructed at the time `pin()` is called. For
-    // that case, we should use `HANDLE.try_with()` instead.
-    HANDLE.with(|handle| handle.pin())
+    with_handle(|handle| handle.pin())
 }
 
 /// Returns `true` if the current thread is pinned.
 #[inline]
 pub fn is_pinned() -> bool {
-    // FIXME(jeehoonkang): thread-local storage may be destructed at the time `pin()` is called. For
-    // that case, we should use `HANDLE.try_with()` instead.
-    HANDLE.with(|handle| handle.is_pinned())
+    with_handle(|handle| handle.is_pinned())
 }
 
 /// Returns the default handle associated with the current thread.
 #[inline]
 pub fn default_handle() -> Handle {
-    HANDLE.with(|handle| handle.clone())
+    with_handle(|handle| handle.clone())
 }
 
 /// Returns the default handle associated with the current thread.
 #[inline]
 pub fn default_collector() -> &'static Collector {
     &COLLECTOR
 }
+
+#[inline]
+fn with_handle<F, R>(mut f: F) -> R
+where
+    F: FnMut(&Handle) -> R,
+{
+    HANDLE.try_with(|h| f(h)).unwrap_or_else(|_| f(&COLLECTOR.register()))
+}
+
+#[cfg(test)]
+mod tests {
+    use crossbeam_utils::thread;
+
+    #[test]
+    fn pin_while_exiting() {
```

Comment:
Is this only meant to not abort? If so a comment explaining that would be useful, now the test doesn't really make much sense.